### PR TITLE
Git LFS support v2

### PR DESCRIPTION
**Pull request by fabian-z**
Requested in #80 

This PR implements Git LFS support for Gitea, supporting the [legacy v1 API](https://github.com/github/git-lfs/blob/master/docs/api/v1/http-v1-legacy.md) as well as the current [batch v1 API](https://github.com/github/git-lfs/blob/master/docs/api/v1/http-v1-batch.md).

Originally imported LFS server code base from https://github.com/git-lfs/lfs-test-server.
Rebased from original PR gogits/gogs#3868 for Gitea, considering inital review in #120.

Features:
* Uses [recommended paths](https://github.com/github/git-lfs/tree/master/docs/api) with HTTP and SSH remotes, no custom user settings necessary
* Supports basic authentication with user credentials
* Supports [authentication with SSH](https://github.com/github/git-lfs/tree/master/docs/api) (passing JWT under the hood)
* Supports no authentication when pulling public repositories ~~(however git-lfs still asks for credentials, where the prompt can be skipped - likely a bug in git-lfs)~~ (Workaround for git-lfs bug applied in https://github.com/go-gitea/gitea/pull/122/commits/d1a2dca05b49b0f9ae8f43e7e8e116f8c42a4a62)
* Settings transparently integrated into installation process
* Removes LFS objects when repositories are deleted. Works with forks (e.g. when parent repository is deleted)

Notes:
* LFS installation setting does not show added locale strings because bindata is not regenerated in this PR (as requested by @strk in #120)

Sorry for my misunderstanding regarding this projects state. Looking forward to working together with you! :smile:

### CODE REVIEW

**Comment by lunny** [comment]
Great!

**Comment by lunny** [comment]
This will resolved #80

File: models/repo.go
Comment (tboerger) [inline models/repo.go]: `if err := sess.Delete(&LFSMetaObject{RepositoryID: repoID}); err != nil {`

**Review by tboerger — CHANGES_REQUESTED** [review]

File: models/repo.go
Comment (fabian-z) [reply models/repo.go]: Changed in https://github.com/go-gitea/gitea/pull/122/commits/a2945dd431edc448f94cadc72c7168bbb2e4cc03

**Comment by tboerger** [comment]
As far as I can say LGTM

**Comment by lunny** [comment]
but how to test it?

**Comment by fabian-z** [comment]
If you are referring to functional testing, I did a fair bit of that with this code as of https://github.com/go-gitea/gitea/pull/122/commits/316b01210e7ac8407d56f9b4dcb6c2fbd6d780a6. Since it uses the recommended HTTP endpoints and should be feature complete, testing actions would be identical to using LFS with GitHub:
HTTP and SSH auth push/pull/clone, fork repository with LFS files and delete parent repo, try pushing LFS files without write access, etc.
GitHub provides some [example usage](https://github.com/github/git-lfs/wiki/Tutorial) to get started.

Regarding unit tests, it may be possible to adapt some of the tests originally found in [lfs-test-server](https://github.com/git-lfs/lfs-test-server) in the long run. Since the module interacts with a lot of subsystems in Gitea, it will be quite challenging to do for a large part of the actual server code. Content store tests should not be as difficult.

File: modules/lfs/server.go
Comment (lunny) [inline modules/lfs/server.go]: `gitea-lfs` :)

File: models/lfs.go
Comment (lunny) [inline models/lfs.go]: Adding an `index` xorm tag to `RepositoryID` and an unique index to `Oid` and `RepositoryID` is better.

File: models/repo.go
Comment (lunny) [inline models/repo.go]: Use `Count` is better.

File: modules/lfs/content_store.go
Comment (lunny) [inline modules/lfs/content_store.go]: `f` is not be closed

File: models/lfs.go
Comment (fabian-z) [reply models/lfs.go]: Not sure what you mean, technically with LFS there's currently only the v1 legacy API, v1 batch API and the v1.3 batch API (which is only a slightly extended version of v1 batch). The LFSMetaObject model should fit all API versions currently specified.

**Comment by fabian-z** [comment]
@lunny 
Thanks for your review! I changed the aspects you pointed out in my commits after https://github.com/go-gitea/gitea/pull/122/commits/6814bdce1f6a881fe65f6503bf26f47c22982608. I had to rebase to test my changes since I run into hook rejected errors on pushing (most likely related to #151).

File: modules/lfs/content_store.go
Comment (fabian-z) [reply modules/lfs/content_store.go]: Good catch! I'll file this as an issue/PR with the lfs-test-server guys.

Comment (fabian-z) [reply modules/lfs/content_store.go]: Turns out the issue was already filed as https://github.com/git-lfs/lfs-test-server/pull/49 with the lfs-test-server project since it makes unit tests on Windows fail. ~~I'm personally more in favor of returning an *os.File handle instead of an io.ReadCloser.~~
EDIT: Since the file is not opened with a read-only flag, it may be more sensible to return an ReadCloser. I changed my last commit accordingly.

File: models/lfs.go
Comment (lunny) [reply models/lfs.go]: I mean maybe we should store the version of the object for a format.

**Comment by lunny** [comment]
@fabian-z Great work! I have tested this according to github's lfs document and it passed. After your change, I will test it again and I wish this PR could be merged ASAP.

**Comment by lunny** [comment]
And I think maybe you could add some information on `runWeb()` functions if `StartLfsServer` is true. i.e. `LFS server is enabled.` I think #151 should have been resolved, if you have any problem, please fire an issue.

File: models/lfs.go
Comment (fabian-z) [reply models/lfs.go]: I don't think there is anything more to store about the object. The object itself does not have any notion of versioning - it's basically just a sha256 hash stored together with the object size. If the tracked file changes, a new meta object will be generated on push (with the new hash).
But maybe I still got you wrong - please don't hesitate to explain your point to me if this is the case :smile:

**Comment by fabian-z** [comment]
Glad to hear it worked well for you!
#151 errors disappeared for me after rebasing, so I think it should be resolved.

I added some informational logging in runWeb() as you suggested, however it is worth noting that this info log will most likely not appear directly after installation, but it will log on subsequent starts of Gitea. This is why I had to make the module stateless in https://github.com/go-gitea/gitea/pull/122/commits/316b01210e7ac8407d56f9b4dcb6c2fbd6d780a6.

If I'm not mistaken I should have incorporated all reviews, so feel free to merge if there are no issues left from your side.

**Comment by lunny** [comment]
Another question, on Gitea UI, we can see something like

```
version https://git-lfs.github.com/spec/v1
oid sha256:5e9676cedcd87dd0909bf7007bfba571354f61d728deb52a426a3be105ba3774
size 110981778
```

but I think maybe it should not be shown to user, user should see `binary` and could download it.

**Comment by lunny** [comment]
see https://github.com/lunny/lfs-test/blob/master/Docker.dmg, when you view a lfs file on UI. It will show `Stored with Git LFS` and you can download it, but not show the meta information. If you can do like github, it's perfect.

**Comment by fabian-z** [comment]
Good idea, I forgot about the UI part initially. I implemented this in https://github.com/go-gitea/gitea/pull/122/commits/2b02b049321ebe595d7170e9a9c371699d78c36e and https://github.com/go-gitea/gitea/pull/122/commits/041748e0a2291d57c3e4cdc32ba72215413caa39. Also, with https://github.com/go-gitea/gitea/pull/122/commits/d1a2dca05b49b0f9ae8f43e7e8e116f8c42a4a62 the module now works around https://github.com/github/git-lfs/issues/1088, so when pulling/cloning public repositories the git-lfs client does not ask for authentication credentials anymore. Also did a trivial rebase to remove a merge conflict in cmd/serve.go.

@lunny If you could crosscheck the functionality of my latest changes - especially regarding authentication, I think we are good to go!

**Comment by fabian-z** [comment]
I just saw that the Travis CI check failed for Go 1.7 with
` Failed to fetch http://downloads-distro.mongodb.org/repo/debian-sysvinit/dists/dist/10gen/i18n/Translation-en  Unable to connect to downloads-distro.mongodb.org:http:`
Since it succeeded for Go 1.6, this check failure is most likely unrelated to the actual tests.

**Review by tboerger — APPROVED** [review]

**Comment by lunny** [comment]
Great work. Only delete the lfs file will return 500 error.

```
 [...uters/repo/editor.go:362 DeleteFilePost()] [E] DeleteRepoFile: UpdateLocalCopyBranch [branch: master]: git clone master: exit status 128 - Cloning into 'data/tmp/local-rpeo/2'...
done.
Downloading Docker.dmg (105.84 MB)
Error downloading object: Docker.dmg (5e9676cedcd87dd0909bf7007bfba571354f61d728deb52a426a3be105ba3774)

Errors logged to /Users/lunny/gopath/src/code.gitea.io/gitea/data/tmp/local-rpeo/2/.git/lfs/objects/logs/20161115T150134.939368495.log
Use `git lfs logs last` to view the log.
error: external filter git-lfs smudge -- %f failed 2
error: external filter git-lfs smudge -- %f failed
fatal: Docker.dmg: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

It seems to ask the server installed lfs plugin?

**Comment by fabian-z** [comment]
I just checked and since Gitea has to use the git command to checkout the repository, it seems to fail if it is running under a user that has installed the git-lfs hooks. Without the hooks it does a regular checkout and deletes the metadata file without any problem.

So I think there are ~~two~~ three possibilities:
1. Document the user needs to run Gitea under a OS user that does not have the LFS hooks installed 
2. Remove the option to delete LFS stored files from the UI
3. Change git invocation (see below)

IMHO the ~~first~~ third option would be the most sensible choice. Without the hooks it will only download the tiny metadata files that are actually stored in the git repository.

**Comment by fabian-z** [comment]
A third option: Make git ignore the filters from command line invocation.
**EDIT**: My last commit wrongly changed the vendored code.gitea.io/git, so I removed it again, see the attached patch below.

This option will require git v2.1.2 as of https://github.com/git/git/commit/a789ca70e7a5b02973b116d21674acd795238f99 and will need a patch in code.gitea.io/git:

```
From 5801cdb3b037395c3c9cd0758fe1088ab9675317 Mon Sep 17 00:00:00 2001
From: Fabian <fabian.zaremba@nf-design.eu>
Date: Tue, 15 Nov 2016 13:22:11 +0100
Subject: [PATCH] Ignore LFS hooks on git command invocation

---
 command.go | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/command.go b/command.go
index f7fce67..dbf3919 100644
--- a/command.go
+++ b/command.go
@@ -28,9 +28,10 @@ func (c *Command) String() string {

 // NewCommand creates and returns a new Git Command based on given command and arguments.
 func NewCommand(args ...string) *Command {
+   ignoreLFS := []string{"-c", "filter.lfs.required=", "-c", "filter.lfs.smudge=", "-c", "filter.lfs.clean="}
    return &Command{
        name: "git",
-       args: args,
+       args: append(ignoreLFS, args...),
    }
 }

-- 
2.10.2


```

**Comment by lunny** [comment]
I think the third method should be OK. But maybe we should add the ignore flags from gitea repo but not change git repo. When delete a file, if it's a lfs file, give a lfs ingore flag to the function to do that?

**Comment by fabian-z** [comment]
Unfortunately, it is not just deleting a file. It may let any git operation involving clone, commit,  etc. fail.
I just tested and it is also not possible to e.g. create a new file or edit a file when the hooks are applied for the Gitea OS user.

I did not find any usage of the `NewCommand` function in Gitea, every invocation seems to be in the `code.gitea.io/git` repository, some I'm not sure how it would be possible to efficiently pass a flag for not just one but many git operations without changing a lot of code.

**Comment by fabian-z** [comment]
To put in my two cents: The `code.gitea.io/git` repository is built for automated interaction with git repositories by using the command-line git client. The git client itself is perfectly suited for that interaction. However, the git-lfs addon client does not play nice with this automated interaction (prompting for authentication, etc.). So it would seem reasonable to me to generally disable the hooks when git is used by this package.

**Comment by lunny** [comment]
Maybe we could just disable delete and edit git lfs file on this PR. And we can open a new PR to do that after we improved `code.gitea.io/git`?

**Comment by fabian-z** [comment]
My testing suggests that we would need to disable all file changing operations for repositories using any LFS tracked files, if compatibility with running Gitea under an OS user having the git-lfs hooks installed is really needed in the short term. It affects at least all git cloning operations, making any change to any repository file impossible over the UI.

But that should be an uncommon use case anyway - best practice is that a different OS user should be used for running a server daemon. I therefore think it would be enough to document this fact as a simple 

> LFS  support currently requires Gitea to be run under an OS user that does not have the git-lfs client hooks installed

and improve `code.gitea.io/git` afterwards in a new PR as you suggested.

**EDIT**: I could also integrate a check based on the output of `git config -l` checking for the `filter.lfs.*` strings and disable LFS support if the hooks are installed, logging an error message if it was configured to start the LFS server.

**Comment by lunny** [comment]
So I have sent a PR to git to add global command args support. After that is merged, you could set the `git.GlobalCommandArgs = append(git.GlobalCommandArgs, "-c", "filter.lfs.required=", "-c", "filter.lfs.smudge=", "-c", "filter.lfs.clean=")` and check `git.BinVersion()` if LFS is enabled.

**Comment by fabian-z** [comment]
Nice, thanks for that! Should I update `vendor/code.gitea.io/git` with govendor in this PR  when incorporating the new check and setting or should that dependency update be a new PR, too?

**Comment by lunny** [comment]
A new pr is better I think

**Comment by fabian-z** [comment]
I will provide a dependency update PR and push a new commit incorporating the needed changes as soon as go-gitea/git#14 is merged.

**Comment by strk** [comment]
I think the vendor/ update, if required for LFS support,
should be part of this PR.

As per #187 (please review)

**Comment by fabian-z** [comment]
So if no one disapproves the dependency update will be integrated by me in a separate commit to this PR branch.

**Comment by lunny** [comment]
@fabian-z it's your turn. :)

**Comment by fabian-z** [comment]
I additionally needed to update some code for the changed package constants in `code.gitea.io/git`, everything is working fine in my tests now.

**Comment by lunny** [comment]
Work fine. But I think this should be the last one. 

After I added or deleted one lfs file, on commits view, when I select the commit, it displayed the lfs' meta informations.

**Comment by fabian-z** [comment]
https://github.com/go-gitea/gitea/pull/122/commits/7954e7faa1102c20cdbd3a2c47f19e321322590f should hide the diff, with https://github.com/go-gitea/gitea/pull/122/commits/64c69057558fbd87218dada8e5585581f365606f adding a "Stored with Git LFS" notice to the file in commit view. Looking forward to your feedback :smile:

**Comment by lunny** [comment]
It seems already ready for merge. But maybe we have to wait after we published v1.0 branch. @tboerger @bkcsoft

**Comment by lunny** [comment]
But there is an build error because another PR merged changed the template name from `INSTALL` to `tplInstall`.

**Comment by fabian-z** [comment]
I had already fixed this with a rebase shortly after pushing my final two commits, odd that you still got this error. Travis also shows as successful for me, where it failed because of the changed template name before my rebase. Could you try again?

**Comment by ghost** [comment]
Hmm... LGTM.

**Comment by lunny** [comment]
This PR will be merged after we publish v1.0.

**Comment by HZ89** [comment]
[Macaron] 2016-12-19 12:28:55: Started GET /root/test.git/info/refs?service=git-receive-pack for 192.168.42.251
2016/12/19 12:28:55 [D] Session ID: c5e929d71282bb77
2016/12/19 12:28:55 [D] CSRF Token: GeD8WYx0Ky8IsNhpZWcQiljaISE6MTQ4MjEyMTczNTE4NTUyOTA5NQ==
[Macaron] 2016-12-19 12:28:55: Completed /root/test.git/info/refs?service=git-receive-pack 401 Unauthorized in 1.708102ms
[Macaron] 2016-12-19 12:28:55: Started GET /root/test.git/info/refs?service=git-receive-pack for 192.168.42.251
2016/12/19 12:28:55 [D] Session ID: 4a67f1e265f4a059
2016/12/19 12:28:55 [D] CSRF Token: n8KMb776eWPhNltVrSRNSjvtox86MTQ4MjEyMTczNTE4NzM5OTg4NA==
[Macaron] 2016-12-19 12:28:55: Completed /root/test.git/info/refs?service=git-receive-pack 200 OK in 103.011545ms
[Macaron] 2016-12-19 12:28:55: Started GET /root/test.git/info/refs?service=git-upload-pack for 192.168.42.251
2016/12/19 12:28:55 [D] Session ID: 2d5604772c5a9a3c
2016/12/19 12:28:55 [D] CSRF Token: JBEJeSiOP0n5Vt0yOWHVLz8m6KY6MTQ4MjEyMTczNTMwMzkzNjkyNA==
[Macaron] 2016-12-19 12:28:55: Completed /root/test.git/info/refs?service=git-upload-pack 401 Unauthorized in 1.780018ms
[Macaron] 2016-12-19 12:29:15: Started GET /root/test/issues for 192.168.42.251
2016/12/19 12:29:15 [D] Session ID: ee2d0933e89eae79
2016/12/19 12:29:15 [D] CSRF Token: PdCI_FybPIG3VUwdvQypZjCCJM06MTQ4MjExNTk0MDIzMDUyMTU5MQ==
------------------------------------------------------------------------------------------------------------------
gitea 1.0.0+dev add a lfs tracked file, but can not push, always return 401

**Comment by lunny** [comment]
This PR has not been merged, so 1.0.0+dev will not support LFS.

**Comment by lunny** [comment]
Please resolve the conflict and we will merge it.

**Comment by fabian-z** [comment]
I fixed the conflicts by rebasing on top of current master and corrected lint errors in my last commit. Also a ```not before``` claim was added to the SSH authentication JWT token, which should improve security.
All LFS features are working fine in my tests. I hope this is ready for merging now and look forward to your final testing and feedback!

**Comment by lunny** [comment]
let LGTM work

**Comment by fabian-z** [comment]
@lunny Thank you for your great review! Nice this got merged as the polished feature it is now.

**Comment by lunny** [comment]
@fabian-z Thanks for your great contribution. :)

**Pull request merged**
